### PR TITLE
[IPv6] Fix checksum calculation of IPv6 with extension headers.

### PR DESCRIPTION
**Pull request by ActoryOu**
[IPv6] Fix checksum calculation of IPv6 with extension headers- RFC 2460 section 4.1

Description
-----------

With this change, FreeRTOS-Plus-TCP supports to calculate checksum for IPv6 packets with IPv6 extension headers.

Test Steps
-----------

Pass IPv6 TCP/UDP/ICMP checksum test cases with this change.

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [x] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: source/FreeRTOS_IPv6_Utils.c [hunk 1]
```diff
@@ -161,6 +172,117 @@ BaseType_t prvChecksumICMPv6Checks( size_t uxBufferLength,
...
+        if( xGetExtensionOrder( ucCurrentHeader, 0U ) > 0 )
+        {
+            while( ( uxIndex + 8U ) < uxBufferLength )
+            {
+                ucNextHeader = pucEthernetBuffer[ uxIndex ];
+
+                xCurrentOrder = xGetExtensionOrder( ucCurrentHeader, ucNextHeader );
+
+                /* To avoid compile warning if debug print is disabled. */
+                ( void ) xCurrentOrder;
```
Comment (tony-josi-aws) [inline source/FreeRTOS_IPv6_Utils.c:215]: This line may not be needed as `xCurrentOrder` is getting used in line 277.

**Comment by tony-josi-aws** [comment]
In the case of NA, NS and RS packets shouldn't we offload the checksum calculation to hardware based on `ipconfigDRIVER_INCLUDED_TX_IP_CHECKSUM` flag?

File: source/FreeRTOS_IPv6_Utils.c (hunk 1 above)
Comment (ActoryOu) [reply source/FreeRTOS_IPv6_Utils.c:215]: No, line 277 is using `ucCurrentHeader`, not `xCurrentOrder`.

**Comment by ActoryOu** [comment]
> In the case of NA, NS and RS packets shouldn't we offload the checksum calculation to hardware based on `ipconfigDRIVER_INCLUDED_TX_IP_CHECKSUM` flag?

Good catch. Updated to check `ipconfigDRIVER_INCLUDED_TX_IP_CHECKSUM` before calling `usGenerateProtocolChecksum` in NA/NS/RS outgoing packets.

**Review by tony-josi-aws — APPROVED** [review]

**Review by tony-josi-aws — APPROVED** [review]

File: source/FreeRTOS_ND.c [hunk 2]
```diff
@@ -639,9 +639,18 @@
                 ( void ) memcpy( pxICMPHeader_IPv6->ucOptionBytes, pxEndPoint->xMACAddress.ucBytes, ipMAC_ADDRESS_LENGTH_BYTES );
 
                 /* Checksums. */
-                pxICMPHeader_IPv6->usChecksum = 0U;
-                /* calculate the ICMP checksum for the outgoing package. */
-                ( void ) usGenerateProtocolChecksum( pxDescriptor->pucEthernetBuffer, pxDescriptor->xDataLength, pdTRUE );
+                #if ( ipconfigDRIVER_INCLUDED_TX_IP_CHECKSUM == 0 )
+                    {
+                        /* calculate the ICMPv6 checksum for outgoing package */
```
Comment (moninom1) [inline source/FreeRTOS_ND.c:644]: Is the pxICMPHeader_IPv6->usChecksum = 0U; not needed here? Earlier we were calling generate checksum after initializing to zero.

Comment (ActoryOu) [reply source/FreeRTOS_ND.c:644]: It's not necessary because we use 0 as initial value when calculating checksum.

**Comment by moninom1** [comment]
Can you please mentioned RFC in the commit message as well.

**Review by AniruddhaKanhere — APPROVED** [review]

**Pull request merged**